File: leetcode/leetcode_py/h-index.py
'''
https://leetcode.com/problems/h-index

Solution1. brute force
time: o(n^2)
space: o(1)

Solution2. sort
time: o(nlogn)
space: o(1)

1. sort
2. iterate from end and find value in index(n-h-1) is at leasst h or not

Solution3. hash table
time: o(n)
space: o(n)
'''
class Solution:
    def hIndex_1(self, citations: List[int]) -> int:
        n = len(citations)
        for i in range(n, 0, -1):
            cnt = 0
            for num in citations:
                if num >= i:
                    cnt += 1
                if cnt == i:
                    return i
        return 0
            
    def hIndex_2(self, citations: List[int]) -> int:
        n = len(citations)
        citations.sort()
        for i in range(n, 0, -1):
            if citations[n-i] >= i:
                return i
        return 0
            
    def hIndex_3(self, citations: List[int]) -> int:
        n = len(citations)
        h = [0] * (n+1)
        for num in citations:
            if num >= n:
                h[n] += 1
            else:
                h[num] += 1
        cnt = 0
        for i in range(n, 0, -1):
            cnt += h[i]
            if cnt >= i:
                return i
        return 0

    def hIndex(self, citations: List[int]) -> int:
        return self.hIndex_3(citations)
        

File: leetcode/leetcode_cpp/h-index.cpp
/*
https://leetcode.com/problems/h-index

Solution1. brute force
time: o(n^2)
space: o(1)

Solution2. sort
time: o(nlogn)
space: o(1)

1. sort
2. iterate from end and find value in index(n-h-1) is at leasst h or not

Solution3. hash table
time: o(n)
space: o(n)

*/
class Solution {
public:
    int hIndex_1(vector<int>& citations) {
        int n = citations.size();
        if (!citations.size()) return 0;
        int res = 0;
        for (int i=n; i>=0; --i) {
            int count = 0;
            for (auto num : citations) {
                if (num >= i)
                    count++;
                if (count == i) {
                    return i;
                }
            }
        }
        return 0;
    }

    int hIndex_2(vector<int>& citations) {
        sort(citations.begin(), citations.end());
        int n = citations.size();
        for (int i=n; i>=1; --i) {
            if (citations[n-i] >= i)
                return i;
        }
        return 0;
    }

    int hIndex_3(vector<int>& citations) {
        int n = citations.size();

        vector<int> hash(n+1, 0);
        for (auto num : citations) {
            if (num >= n)
                hash[n]++;
            else
                hash[num]++;
        }

        int count = 0;
        for (int i=n; i>=1; --i)
        {
            count += hash[i];
            if (count >= i)
                return i;
        }

        return 0;
    }


    int hIndex(vector<int>& citations) {
        return hIndex_3(citations);
    }
};